Bluetooth: hci: Do not drop hci buffers

jori-nordic opened the pull request:
This fixes two logic problems:

HCI rpmsg driver: if we fail to get a buffer immediately, we drop the
hci buffer.

hci_rpmsg sample: if ipc_service_send() times out waiting for a buffer,
we drop the buffer instead of retrying.

We now wait in the IPC RX endpoint, but this shouldn't be a problem
with PR #43806 that will introduce separate workqueues per instance,
with configurable priorities.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jhedberg commented:
Could you fix the commit prefix to be more specific, since this is not about HCI in general, but about the rpmsg transport in particular. E.g. "Bluetooth: hci: rpmsg: ...". I think it'd also be prudent to split this up into two commits.

jhedberg reviewed (changes requested):
See my earlier comment

cvinayak reviewed (changes requested):
My preferred commit title: `samples: Bluetooth: hci_rpmsg: Fix event buffer handling", and mention the enhancement/issue that is being fixed in this PR.

The changes relate to samples and not Bluetooth subsystem.

jori-nordic commented:
> My preferred commit title: `samples: Bluetooth: hci_rpmsg: Fix event buffer handling", and mention the enhancement/issue that is being fixed in this PR.
> 
> The changes relate to samples and not Bluetooth subsystem.

So split it two commits right ? Because the changes do in fact affect the bluetooth subsystem: I'm also changing the rpmsg HCI driver's behavior. Or is the drivers/ folder considered not part of the bluetooth subsystem ? In that case I would put `drivers:` as heading for the first commit.

jori-nordic commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -78,14 +78,16 @@ static struct net_buf *bt_rpmsg_evt_recv(const uint8_t *data, size_t remaining)
 	BT_DBG("len %u", hdr.len);
 
-	buf = bt_buf_get_evt(hdr.evt, discardable, K_NO_WAIT);
+	buf = bt_buf_get_evt(hdr.evt, discardable,
+			     discardable ? K_NO_WAIT : K_FOREVER);
+
```
I can't really return anything meaningful here: the caller can't return anything since it's an IPC callback (workqueue context), and returning without doing anything means the buffer is dropped. So I will loop through with 10 seconds as you suggest, having a warning on each timeout.

jori-nordic commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -218,11 +224,14 @@ static int hci_rpmsg_send(struct net_buf *buf)
 
 	LOG_HEXDUMP_DBG(buf->data, buf->len, "Final HCI buffer:");
-	ipc_service_send(&hci_ept, buf->data, buf->len);
+
+	do {
+		err = ipc_service_send(&hci_ept, buf->data, buf->len);
+	} while (err == RPMSG_ERR_NO_BUFF);
```
Yes I think it would spin without yielding. During debugging, I just added a `k_msleep(10)` when failing to send, but removed it before the submission.
I did not think this was a problem (except maybe for logging ?) since by default main thread has a prio of 0, which is pre-emptable by coop threads.
Would you suggest having a `k_yield` call when the error code is != 0 ? That'd be ok with me.

I'll also add a warning after a reasonable timeout: the openamp docs are currently wrong (well I think it's the timing calculation that is) and the timeout is not 15 seconds, but rather 150ms. So I think not a warning every failed allocation, that'd be too much noise, but a warning every second would be okay. Do you agree ?

jori-nordic committed:
Bluetooth: hci: rpmsg: do not drop hci buffers

This fixes an issue where we would drop the hci event if allocation from
the hci event buffer pool didn't immediately succeed.

The behavior is now to block on allocation, and warn the user every 10
seconds.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
samples: Bluetooth: hci_rpmsg: Fix event buffer handling

This fixes an issue where we dropped the send buffer when
ipc_service_send() timed out waiting for a shared memory buffer.

We now wait in the IPC RX endpoint, but this shouldn't be a problem for
other IPC users, as PR #43806 introduces separate workqueues
per instance, with configurable priorities.

This also includes small changes regarding the IPC service return value.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented:
@cvinayak @jhedberg could you review again ? I split the commits and applied the proposed changes.
